src/Widgets/PartitionMenu.vala

```diff
@@ -0,0 +1,355 @@
+        bottom_revealer.add (bottom_grid);
+
+        grid.attach (top_controls,    0, 0);
+        grid.attach (bottom_revealer, 0, 1);
+
+        add (grid);
+        grid.show_all ();
+
+        // FIXME: Probably should stick these in a revealer, too
+        custom.visible = false;
+        custom_label.visible = false;
+
+        format_partition.notify["active"].connect (() => {
+            if (!disable_signals) {
+                check_values (set_mount);
+            }
+        });
+
+        use_as.changed.connect (() => {
+            if (disable_signals) {
+                return;
+            }
+
+            var active = use_as.active;
+            bool visible = active == 4;
+
+            custom.visible = visible;
+            custom_label.visible = visible;
+
+            if (active == 2) {
+                if (Distinst.bootloader_detect () == Distinst.PartitionTable.GPT) {
+                    type.active = 2;
+                } else {
+                    type.active = 0;
+                }
+                type_label.visible = true;
+                type.visible = true;
+                type.sensitive = false;
+                format_label.visible = true;
+                format_partition.visible = true;
+            } else if (active == 3) {
+                format_label.visible = false;
+                format_partition.visible = false;
+                disable_signals = true;
+                format_partition.active = true;
+                disable_signals = false;
+                type_label.visible = false;
+                type.visible = false;
+            } else {
+                type_label.visible = true;
+                type.visible = true;
+                type.sensitive = true;
+                format_label.visible = true;
+                format_partition.visible = true;
+            }
+
+            check_values (set_mount);
+        });
+
+        type.changed.connect (() => {
+            if (!disable_signals) {
+                check_values (set_mount);
+                set_format_sensitivity ();
+            }
+        });
+
+        custom.changed.connect (() => {
+            if (!disable_signals) {
+                check_values (set_mount);
+            }
+        });
+
+        use_partition.notify["active"].connect (() => {
+            if (disable_signals) {
+                return;
+            }
+
+            if (use_partition.active) {
+                disable_signals = true;
+                set_format_sensitivity ();
+                disable_signals = false;
+                use_as.set_active (
+                    (fs == Distinst.FileSystemType.FAT16 || fs == Distinst.FileSystemType.FAT32)
+                        ? mount_set (boot_partition) ? 4 : 2
+                        : fs == Distinst.FileSystemType.SWAP
+                            ? 3
+                            : mount_set ("/")
+                                ? mount_set ("/home" ) ? 4 : 1
+                                : 0
```

Comment:
I'm not a fan of this kind of syntax at all, it's the role of the compiler to optimize it, please use temporary variables